src/scheduler.rs

```diff
@@ -259,25 +274,45 @@ async fn fetch_idle(ctx: &Context, connection: &mut Imap, folder_config: Config)
 
     // Synchronize Seen flags.
     connection
-        .sync_seen_flags(ctx, &watch_folder)
+        .sync_seen_flags(context, &watch_folder)
         .await
         .context("sync_seen_flags")
-        .ok_or_log(ctx);
-
-    connection.connectivity.set_connected(ctx).await;
-
-    // idle
-    if !connection.can_idle() {
-        return connection.fake_idle(ctx, Some(watch_folder)).await;
-    }
-
-    match connection.idle(ctx, Some(watch_folder)).await {
-        Ok(v) => v,
-        Err(err) => {
-            connection.trigger_reconnect(ctx).await;
-            warn!(ctx, "{:#}", err);
-            InterruptInfo::new(false)
+        .ok_or_log(context);
+
+    connection.connectivity.set_connected(context).await;
+
+    if let Some(session) = connection.session.take() {
+        if session.can_idle() {
+            info!(context, "IMAP session supports IDLE, using it.");
+            match session
+                .idle(
+                    context,
+                    connection.idle_interrupt_receiver.clone(),
+                    Some(watch_folder),
+                )
+                .await
+                .context("idle")
+            {
+                Ok((session, info)) => {
+                    connection.session = Some(session);
+                    info
```

Comment:
Up2u but i'd prefer explicit return-s for long functions (otherwise a reader should look down through all "}" to understand the purpose of the expr). And in that case `connection.fake_idle(context, Some(watch_folder)).await` lines could be joined (moved outside of if/else branches)